### PR TITLE
Delete choice user case #42

### PR DESCRIPTION
**Pull request by valentinbujdoso**

### CODE REVIEW

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteValidationTest.java [hunk 1]
```diff
@@ -58,8 +58,7 @@ public void invalid_adminKey_is_rejected() throws ReportedException {
 	public void proper_voteId_choiceId_and_adminKey_does_delete_choice() throws ReportedException {
 		String choiceId = voteManager.addChoice(adminInfo.adminKey, adminInfo.voteId, "choice1", "user");
 		String voteId = adminInfo.voteId;
-		String result = voteManager.deleteChoice(adminInfo.voteId, choiceId, adminInfo.adminKey);
-		assertEquals("OK", result);
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteValidationTest.java]: deleteChoice returns "OK"
https://adadocs.demokracia.rulez.org/PDEngine/edemo/master/index.html#3044c21c-7606-4098-8176-9dd5b93c20ad
What is that throw below?

File: src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java [hunk 2]
```diff
@@ -25,7 +25,7 @@ VoteAdminInfo createVote(String voteName, Set<String> neededAssurances, Set<Stri
 
 	String addChoice(String adminKey, String voteId, String choiceName, String user);
 	
-	String deleteChoice(String voteId, String choiceId, String adminKey) throws ReportedException;
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java]: https://adadocs.demokracia.rulez.org/PDEngine/edemo/master/index.html#3044c21c-7606-4098-8176-9dd5b93c20ad

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	public ChoiceDeleteAdminKeyIsUser() {
+		super();
+	}
+
+	@tested_feature("Manage votes")
+	@tested_operation("delete choice")
+	@tested_behaviour("if \"user\" is used as adminKey, then the user must be the one who added the choice and canAddIn be true")
+	@Test
+	public void adminKey_is_user_and_the_user_is_the_same_with_that_user_who_added_the_choice_and_canAddin_is_false() throws ReportedException {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java]: if canAddin is false then other users cannot add choices

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		assertThrows(
+			() -> voteManager.deleteChoice(adminInfo.voteId, choiceId , "user")
+		).assertMessageIs("The adminKey is \"user\" but canAddin is false.");
+	}
+	
+	@tested_feature("Manage votes")
+	@tested_operation("delete choice")
+	@tested_behaviour("if \"user\" is used as adminKey, then the user must be the one who added the choice and canAddIn be true")
+	@Test
+	public void adminKey_is_user_and_the_user_is_not_the_same_with_that_user_who_added_the_choice_and_canAddin_is_true() throws ReportedException {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java]: if adminKey is user and the user is not the one who added the choice then the choice cannot bedeleted

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+			() -> voteManager.deleteChoice(adminInfo.voteId, choiceId , "user")
+		).assertMessageIs("The adminKey is \"user\" but the user is not same with that user who added the choice.");
+	}
+	
+	
+	@tested_feature("Manage votes")
+	@tested_operation("delete choice")
+	@tested_behaviour("if \"user\" is used as adminKey, then the user must be the one who added the choice and canAddIn be true")
+	@Test
+	public void adminKey_is_user_and_the_user_is_the_same_with_that_user_who_added_the_choice_and_canAddin_is_true() throws ReportedException {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java]: if adminKey is user and canAddin is true then the user who added the choice is able to delete it

**Review by magwas — CHANGES_REQUESTED** [review]

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteValidationTest.java (hunk 1 above)
Comment (magwas) [reply src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteValidationTest.java]: Ok, I understand the throw now.

File: src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java (hunk 2 above)
Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java]: The modify choice has the same in the Deviaton Model (I mean the return "OK") but it is void in this interface.
So which is the good?

Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java]: The modify choice has the same in the Deviaton Model (I mean the return "OK") but it is void in this interface.
So which is the good?

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java (hunk 5 above)
Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java]: Fixed

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java (hunk 4 above)
Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java]: Fixed

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java (hunk 3 above)
Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/ChoiceDeleteAdminKeyIsUser.java]: Fixed

File: src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java (hunk 2 above)
Comment (magwas) [reply src/main/java/org/rulez/demokracia/pdengine/IVoteManager.java]: This diff made it void. It should be String.

**Review by magwas — APPROVED** [review]
nice

**Comment by magwas** [comment]
testcasecount=3

File: src/main/java/org/rulez/demokracia/pdengine/VoteRegistry.java [hunk 6]
```diff
@@ -64,8 +64,17 @@ public String deleteChoice(String voteId, String choiceId, String adminKey) thro
 		Vote vote = getVote(voteId);
 		vote.checkAdminKey(adminKey);
 		Choice votesChoice = vote.getChoice(choiceId);
-		vote.choices.remove(votesChoice.id);
 		
+		if(adminKey.equals("user"))
+			if(votesChoice.userName.equals(getWsUserName()))
+				if(vote.canAddin)  
+					vote.choices.remove(votesChoice.id);
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/VoteRegistry.java:71]: uncovered code

**Review by magwas — CHANGES_REQUESTED** [review]

File: src/main/java/org/rulez/demokracia/pdengine/VoteRegistry.java (hunk 6 above)
Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/VoteRegistry.java:71]: Strange. I missed something on the tests.

Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/VoteRegistry.java:71]: Fixed

**Review by magwas — APPROVED** [review]

**Pull request merged**